### ISSUE TITLE
Add platform-specific window code

### ISSUE THREAD

**Issue by ashkitten** [tomaka/winit#257]
What I'm proposing is that platform-specific window code be added to winit, such as override-redirect, setting Atoms, etc. on Xorg. This would be very useful to my current project, and allow it to be implemented in a much less hacky way. I'd be willing to write a PR for this if you would be willing to potentially merge it.

**Comment by tomaka** [comment]
The API design already exists: https://github.com/tomaka/winit/blob/master/src/os/unix.rs
What may need to be discussed is the actual platform-specific features to expose.

**Comment by ashkitten** [comment]
I think we should expose override-redirect at the very least, since that would only require changing a few lines of code. We could also provide a way to create Atoms, like setting `_NET_WM_WINDOW_TYPE = _NET_WM_WINDOW_TYPE_DESKTOP`. The other thing I need personally is XLowerWindow, but I'm sure other people have other things they'd like to see.

**Comment by jwilm** [comment]
For Alacritty, we use the following platform-specific APIs
* XSetWMProperties - causes WM_CLIENT_MACHINE to be set which is required to set _NET_WM_PID. This is actually used to set a number of properties including WM_CLASS. It's nice to be able to call this before mapping the window so WMs can apply rules immediately.
* XInternAtom - required for creating atoms
* XChangeProperty - we're only using this to set _NET_WM_PID; it uses an Atom returned from XInternAtom

**Comment by Ralith** [comment]
I'm not sure it's appropriate for winit to serve as a low-level platform-specific API binding. I think we already expose all the handles necessary to do whatever stuff you want yourself, and handling everything that people might commonly want could be a very large project.

**Comment by francesca64** [comment]
Long term, `platform::linux::x11::util` could be extracted into its own library, but that's not something I'm interested in doing any time soon. I don't believe it's appropriate for winit to expose any general-purpose X11 abstractions; besides the more obvious aspects of maintaining winit's API surface area and concept, X11 is footgun-heavy, and I'd rather people just request/PR any window configuration stuff they need.

@ashkitten do you still need this? Anything that can be reasonably abstracted via `WindowBuilderExt` can generally be added without much hassle.

@jwilm in the near future I plan on improving the situation with hints on X11, specifically aiming for ICCCM+EWMH compliance and removing the need for directly calling Xlib stuff in Alacritty.

**Comment by ashkitten** [comment]
I would still love to have this, if it would allow me to set properties on windows before they are mapped (right now I have to hackily create the window, set override-redirect, and unmap/remap it)

**Comment by francesca64** [comment]
@ashkitten is override-redirect all you need? I can likely get that done before the next release.

**Comment by ashkitten** [comment]
I'd prefer to be able to set Atoms such as `_NET_WM_WINDOW_TYPE` but override-redirect is the main thing bugging me.

**Comment by francesca64** [comment]
Is this good? #529

**Comment by francesca64** [comment]
Okay, so apparently both `EventsLoopExt` and `WindowExt` have a method called `get_xlib_xconnection`. it's `#[doc(hidden)]` on `EventsLoopExt`, but not on `WindowExt` (I imagine this discrepancy is unintentional, and it should probably always be hidden). Anyway, the reason this matters is because I just merged #534, which means I've unintentionally given you (undocumented) access to a bunch of X11 utils. Even if I hadn't done that, the entire X11 module has been secretly exposed here for a long time: https://github.com/tomaka/winit/blob/30f798b24608aaee9b727f40fb051ad66f445920/src/os/unix.rs#L15-L17

This means you can do stuff like this:

```rust
use winit::os::unix::WindowExt;
use winit::os::unix::x11::util;
let xconn = window.get_xlib_xconnection().unwrap();
let prop_atom = unsafe { xconn.get_atom_unchecked(b"CUTEST_CONTRIBUTOR\0") };
let type_atom = unsafe { xconn.get_atom_unchecked(b"UTF8_STRING\0") };
xconn.change_property(
    window.get_xlib_window().unwrap(),
    prop_atom,
    type_atom,
    util::PropMode::Replace,
    b"francesca64\0",
).flush().expect("Failed to set `CUTEST_CONTRIBUTOR` property");
```
While this likely goes without saying, this API is unstable and unsupported, so use at your own risk.
